Menu: Add install mod button

octacian opened the pull request:
Add install mod button in mods tab on the main menu. Allows installing a mod from ZIP on the local filesystem. Uses built-in file picker and adds a dialog to display whether the installation was successful (will be unsuccessful if the selected file is either not a ZIP file or not a mod).

As you can see from the first screenshot, the Install Mod button is placed beneath the Installed Mods list and the Uninstall selected mod button is moved so as to be placed beside the new button. Due to the size of the resulting buttons, tooltips are used to display more in-depth information. However, if no mod is selected (and therefore the uninstall button is not shown), the install mod button takes full width (seen at bottom of first screenshot).

__Todo:__
- [ ] Support modpacks
- [ ] Improve formspecs
  - [x] Mods tab formspec
  - [ ] Result formspec
- [ ] Support installing from folders

__Screenshots:__ (cropped)
![mt_install_both](https://cloud.githubusercontent.com/assets/15371866/25298015/716d39dc-26a6-11e7-9a5f-05ac848d1e39.png)
![mt_install_filepicker](https://cloud.githubusercontent.com/assets/15371866/25260411/d49522f0-2600-11e7-990d-b137d1fdbde1.png)
![mt_install_success](https://cloud.githubusercontent.com/assets/15371866/25260412/d6229a12-2600-11e7-9a2c-c8e93f3d75b4.png)

sofar commented:
The forms could use some love...

The "install mod" button is really small and awkwardly placed. The symmetry and layout of that form is a small disaster now.

In the file chooser the buttons should be in the bottom I think.

The successful message could write the name of the mod installed on the screen. I don't think it needs to be this large either.

octacian commented:
Overall, I agree.

I only placed the button where I did because I didn't have any other ideas for placement, and yes, I agree that it should be larger.

The file picker is way out of the scope of this PR, it is trigger by the following code which is either defined something else in builtin (outside main menu) or is somewhere in the C++ area (more likely).

Displaying the mod name would require `modmgr.installmod` to return a second parameter so I didn't think to implement this right away, however, I will do so. And yes, your right, it could be smaller. The only reason it is this size is because many other like dialogs are.

rubenwardy commented on builtin/mainmenu/modmgr.lua:
```diff
@@ -395,6 +395,11 @@ function modmgr.installmod(modfilename,basename)
 	modmgr.refresh_globals()
 
+	if basefolder.type == "invalid" then
+		return
```
if this can return false rather than nil, then this if-else can be replaced with:

```lua
return basefolder.type ~== "invalid"
```

rubenwardy commented on builtin/mainmenu/tab_mods.lua:
```diff
@@ -31,16 +31,9 @@ local function get_formspec(tabview, name, tabdata)
-		""
+		";" .. tabdata.selected_mod .. "]" ..
+		"button[10.2,0;1.8,0.5;btn_mod_mgr_install_local;".. fgettext("Install mod") .. "]" ..
+		"tooltip[btn_mod_mgr_install_local;Install mod from ZIP]"
```
needs `fgettext`

rubenwardy commented on builtin/mainmenu/dlg_install_mod.lua:
```diff
@@ -0,0 +1,49 @@
+	local retval =
+		"size[11.5,4.5,true]" ..
+		"label[2,2;" ..
+		fgettext(dialogdata.text) .. "]"..
```
`core.formspec_escape`

octacian replied:
`fgettext` already does this. https://github.com/minetest/minetest/blob/master/doc/menu_lua_api.txt#L201

rubenwardy commented on builtin/mainmenu/tab_mods.lua:
```diff
@@ -31,16 +31,9 @@ local function get_formspec(tabview, name, tabdata)
---		"button[0,4.85;5.25,0.5;btn_modstore;".. fgettext("Online mod repository") .. "]"
-		""
+		";" .. tabdata.selected_mod .. "]" ..
+		"button[10.2,0;1.8,0.5;btn_mod_mgr_install_local;".. fgettext("Install mod") .. "]" ..
```
needs `core.formspec_escape`   (unless fgettext does this already, ie: the f stands for formspec)

octacian replied:
`fgettext` already does this AFAIK (`core.formspec_escape` isn't used _anywhere_ in the main menu but rather fgettext so I assume it does).

rubenwardy commented on builtin/mainmenu/tab_mods.lua:
```diff
@@ -168,7 +157,15 @@ local function handle_buttons(tabview, fields, tabname, tabdata)
 
 	if fields["mod_mgt_open_dlg_accepted"] ~= nil and
 		fields["mod_mgt_open_dlg_accepted"] ~= "" then
```
if statement continuations are doubly indented

octacian replied:
Ah yes, you're right. I've now reread the style guide :D

rubenwardy commented on builtin/mainmenu/dlg_install_mod.lua:
```diff
@@ -0,0 +1,49 @@
+--Minetest
+--Copyright (C) 2014 sapier
```
if you wrote this then you can place your name here

octacian commented:
Fixed all issues mentioned by @rubenwardy.

octacian commented:
![mt_install_mod](https://cloud.githubusercontent.com/assets/15371866/25281908/bbf7fb10-2663-11e7-9f93-e9cd3789f007.png)

Does that look a bit better? I've moved the button to underneath the Installed Mods list where it makes a lot more sense as an install mod button. I've also removed the tooltip and just placed the full text on the button seeing as the new placement allows for more space.

This is just testing though, I haven't yet pushed this change.

Fixer-007 commented:
But what if I want to install NOT ZIPped mod, say it is just a folder.

nerzhul commented:
do you check if init.lua is present in zip before validating ? it's very important to check that and refuse if not

octacian commented:
@Fixer-007 ATM plain folders are not supported, however, this should be completed before this PR is merged, plus, I've found some other issues.

@nerzhul Yes, we do, which brings us to a new problem.

This PR uses `modmgr.installmod` to install mods, which was created before modpacks became a thing. So, before this PR is merged, I'll have to write some code to further check for `modpack.txt` if basic mod-related tests fail.

bigfoot547 commented:
![d49522f0-2600-11e7-990d-b137d1fdbde1](https://cloud.githubusercontent.com/assets/21959175/25310111/41139392-27a2-11e7-877f-4744d24f2c2b.png)
So this is your hacked client?

octacian commented:
> So this is your hacked client?

Yep, that'd be it. More specifically, it doesn't have sneak changes :P 
/me should really pay closer attention to what shows on screenshots :rotfl:

bigfoot547 commented:
It's okay. I have a hacked client too.

red-001 commented:
has #1020 been fixed?

rubenwardy commented:
>which was created before modpacks became a thing. 

No, modpack have been a thing for ages. Mmdb just didn't allow them

GreenXenith commented:
There are lots of mods that use .targz and some that use .7z. Will there be support for those (preferably the .targz)?

nerzhul commented:
tar gz and zip are easier to integrate than 7zip which is less standard (lzma)

octacian commented:
I've decided that I'd like to just entirely rewrite the `modmgr`, including support for client mods and just in general cleaning stuff up. Unless someone objects, I think I will close this PR and submit a new PR with `modmgr` rewrite, following by separate PRs to implement the install mod button and to somehow add support for enabling/disabling and installing/uninstalling client mods through `modmgr` rewrite. It might be better to add a separate tab for this, but either way, `modmgr` rewrite should be a different PR.

With that in mind, I'm going to close this PR for now. If someone feels as though this is not the way I should go about things, feel free to reopen.

rubenwardy commented:
@octacian sounds good!

I hope that you do it in a way that makes modmgr pure, ie: no formspec logic inside it
Would be nice to make it testable too
